### PR TITLE
NEX-3/Registration Pull Request

### PR DESCRIPTION
**Pull request by LuckyOx04**
A pull request for Nex-3/Registration branch to be merged into the dev branch.

### CODE REVIEW

File: nexus/src/main/java/com/example/nexus/exception/GlobalExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    @ExceptionHandler(UserAlreadyExistsException.class)
+    public ResponseEntity<String> handle(UserAlreadyExistsException ex) {
+        log.error(ex.getMessage());
+        log.info(ex.getMessage(), ex);
+
+        return ResponseEntity.status(HttpStatus.CONFLICT).body(ex.getMessage());
+    }
+
+    @ExceptionHandler(UnauthorizedException.class)
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/exception/GlobalExceptionHandler.java]: Combine this handler with AuthenticationExceptio. They both have same structure

File: nexus/src/main/java/com/example/nexus/exception/UnauthorizedException.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.example.nexus.exception;
+
+public class UnauthorizedException extends RuntimeException {
+
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/exception/UnauthorizedException.java]: remove blank line

File: nexus/src/main/java/com/example/nexus/exception/NotFoundException.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.example.nexus.exception;
+
+public class NotFoundException extends RuntimeException {
+
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/exception/NotFoundException.java]: remove blank line

File: nexus/src/main/java/com/example/nexus/constant/RegexConstants.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package com.example.nexus.constant;
+
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/constant/RegexConstants.java:2]: Add  @NoArgsConstructor(access = AccessLevel.PRIVATE)

File: nexus/src/main/java/com/example/nexus/model/payload/request/RegisterRequest.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package com.example.nexus.model.payload.request;
+
+public record RegisterRequest(String firstName, String lastName, Float balance,
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/model/payload/request/RegisterRequest.java]: Remove balance it defautly set to 0

File: nexus/src/main/java/com/example/nexus/service/RegisterServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    private final ProfileRepository profileRepository;
+    private final RoleRepository roleRepository;
+    private final RegisterMapper registerMapper;
+
+    @Override
+    public Profile registerUser(RegisterRequest registerRequest) {
+
+        validatePassword(registerRequest);
+
+        var newUser = registerMapper.mapUser(registerRequest);
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/service/RegisterServiceImpl.java]: use final

File: nexus/src/main/java/com/example/nexus/service/RegisterServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        isUserNameTaken(newUser.getUsername());
+        newUser.setPassword(passwordEncoder.encode(registerRequest.password()));
+        roleRepository.findByName(RoleConstants.USER).
+                ifPresentOrElse(userRole -> newUser.getRoles().add(userRole),
+                        () -> {throw new NotFoundException(MessageConstants.ROLE_NOT_FOUNT);});
+
+        var newProfile = registerMapper.mapProfile(registerRequest);
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/service/RegisterServiceImpl.java]: use final

File: nexus/src/main/java/com/example/nexus/controller/AuthController.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+        return ResponseEntity
+                .status(HttpStatus.CREATED)
+                .headers(headers)
+                .build();
+    }
+
+    @PostMapping("/registration")
+    ResponseEntity<?> register(@RequestBody RegisterRequest registerRequest) {
+        var bodyContent = registerService.registerUser(registerRequest);
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/controller/AuthController.java]: Use final

File: nexus/src/main/java/com/example/nexus/service/RegisterService.java [hunk 9]
```diff
@@ -0,0 +1,8 @@
+package com.example.nexus.service;
+
+import com.example.nexus.model.entity.Profile;
+import com.example.nexus.model.payload.request.RegisterRequest;
+
+public interface RegisterService {
+    Profile registerUser(RegisterRequest registerRequest);
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/service/RegisterService.java]: return some type of Response

File: nexus/src/main/java/com/example/nexus/controller/AuthController.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        headers.setBearerAuth(token);
+
+        return ResponseEntity
+                .status(HttpStatus.CREATED)
+                .headers(headers)
+                .build();
+    }
+
+    @PostMapping("/registration")
+    ResponseEntity<?> register(@RequestBody RegisterRequest registerRequest) {
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/controller/AuthController.java:37]: When your ResponseEntity have body dont use ?. Use type of Response.

**Review by stodorichkov — CHANGES_REQUESTED** [review]

File: nexus/src/main/java/com/example/nexus/exception/GlobalExceptionHandler.java (hunk 1 above)
Comment (LuckyOx04) [reply nexus/src/main/java/com/example/nexus/exception/GlobalExceptionHandler.java]: Done.

File: nexus/src/main/java/com/example/nexus/exception/UnauthorizedException.java (hunk 2 above)
Comment (LuckyOx04) [reply nexus/src/main/java/com/example/nexus/exception/UnauthorizedException.java]: Done.

File: nexus/src/main/java/com/example/nexus/exception/NotFoundException.java (hunk 3 above)
Comment (LuckyOx04) [reply nexus/src/main/java/com/example/nexus/exception/NotFoundException.java]: Done.

File: nexus/src/main/java/com/example/nexus/constant/RegexConstants.java (hunk 4 above)
Comment (LuckyOx04) [reply nexus/src/main/java/com/example/nexus/constant/RegexConstants.java:2]: Done.

File: nexus/src/main/java/com/example/nexus/model/payload/request/RegisterRequest.java (hunk 5 above)
Comment (LuckyOx04) [reply nexus/src/main/java/com/example/nexus/model/payload/request/RegisterRequest.java]: Done.

File: nexus/src/main/java/com/example/nexus/service/RegisterServiceImpl.java (hunk 6 above)
Comment (LuckyOx04) [reply nexus/src/main/java/com/example/nexus/service/RegisterServiceImpl.java]: Done.

File: nexus/src/main/java/com/example/nexus/service/RegisterServiceImpl.java (hunk 7 above)
Comment (LuckyOx04) [reply nexus/src/main/java/com/example/nexus/service/RegisterServiceImpl.java]: Done.

File: nexus/src/main/java/com/example/nexus/controller/AuthController.java (hunk 8 above)
Comment (LuckyOx04) [reply nexus/src/main/java/com/example/nexus/controller/AuthController.java]: Done.

File: nexus/src/main/java/com/example/nexus/service/RegisterService.java (hunk 9 above)
Comment (LuckyOx04) [reply nexus/src/main/java/com/example/nexus/service/RegisterService.java]: Done.

File: nexus/src/main/java/com/example/nexus/controller/AuthController.java (hunk 10 above)
Comment (LuckyOx04) [reply nexus/src/main/java/com/example/nexus/controller/AuthController.java:37]: Done.

File: nexus/src/main/java/com/example/nexus/service/AuthService.java [hunk 11]
```diff
@@ -1,7 +1,11 @@
...
 
+import com.example.nexus.model.entity.Profile;
 import com.example.nexus.model.payload.request.AuthenticationRequest;
+import com.example.nexus.model.payload.request.RegisterRequest;
+import org.springframework.http.ResponseEntity;
 
 public interface AuthService {
     String login(AuthenticationRequest request);
-}
+    ResponseEntity<Profile> registerUser(RegisterRequest registerRequest);
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/service/AuthService.java]: Return Profile or some kind of Response. Not ResponseEntity, I think is better to return some kind of message.

File: nexus/src/main/java/com/example/nexus/service/AuthServiceImpl.java [hunk 12]
```diff
@@ -37,10 +51,45 @@ public String login(AuthenticationRequest request) {
         return "token";
     }
 
+    @Override
+    public ResponseEntity<Profile> registerUser(RegisterRequest registerRequest) {
+
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/service/AuthServiceImpl.java:56]: remove blank line. Also first check for user already exist. Second check passwor validation. And finaly save new user.

File: nexus/src/main/java/com/example/nexus/service/AuthServiceImpl.java [hunk 13]
```diff
@@ -37,10 +51,45 @@ public String login(AuthenticationRequest request) {
...
+        if (!registerRequest.password().matches(RegexConstants.PASSWORD_REQUIREMENT_REGEX)) {
+            throw new UnauthorizedException(MessageConstants.WRONG_PASSWORD_FORMAT);
+        }
+        if (!registerRequest.password().equals(registerRequest.confirmPassword())) {
+            throw new UnauthorizedException(MessageConstants.CONFIRM_PASSWORD_NOT_MATCHING);
+        }
+    }
+
+    private void isUserNameTaken(String username) {
+        userRepository.findByUsername(username).ifPresent(
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/service/AuthServiceImpl.java]: Better use if

**Review by stodorichkov — CHANGES_REQUESTED** [review]

File: nexus/src/main/java/com/example/nexus/controller/AuthController.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        headers.setBearerAuth(token);
+
+        return ResponseEntity
+                .status(HttpStatus.CREATED)
+                .headers(headers)
+                .build();
+    }
+
+    @PostMapping("/registration")
+    ResponseEntity<Profile> register(@RequestBody RegisterRequest registerRequest) {
```
Comment (stodorichkov) [inline nexus/src/main/java/com/example/nexus/controller/AuthController.java]: ResponseEntity<?>

**Review by stodorichkov — CHANGES_REQUESTED** [review]

**Review by stodorichkov — APPROVED** [review]
Everything is 👍

**Pull request merged**